keps/sig-architecture/3935-oldest-node-newest-control-plane/README.md

```diff
@@ -0,0 +1,796 @@
+* This does not address third-party components that version themselves with node components and limit skew relative to the control plane.
+  Maintainers of those components can choose to support larger or smaller version skew. Node operators should limit themselves to skew
+  supported by core Kubernetes node components and any third-party components required for node operation.
+
+## Proposal
+
+1. Add a configuration to https://testgrid.k8s.io/sig-node-kubelet#Summary to begin testing 3 version node / control plane skew
+2. Identify in-flight feature enablement plans from SIGs that rely on a maximum of 2 version node / control plane skew, and the node version when those plans would be complete
+3. Choose the minimum node version to begin supporting 3-version skew that avoids disrupting the plans identified in step 2
+   * 1.25 nodes / 1.28 control planes
+4. Update the [Kubernetes version skew policy](https://kubernetes.io/releases/version-skew-policy/) as part of the release identified in step 3
+
+### Risks and Mitigations
+
+`kubeadm` currently limits itself to managing nodes a single minor version older than the `kubeadm` build.
+This restriction is overridden/ignored when `kubeadm` is used to create clusters in the existing skew CI jobs.
+If the kubelet command-line or configuration file changed in ways `kubeadm` has to adapt to,
+current builds of `kubeadm` could stop being able to successfully configure n-2 or n-3 nodes.
+Note that this risk already exists for n-2 nodes, and has not proved problematic so far.
+
+Possible mitigations include:
+* sig-node extending rollout of command-line or configuration file changes to ease adoption by `kubeadm` over multiple versions.
+  This approach has the benefit of also easing adoption by kubelet end users.
+* sig-cluster-lifecycle restructuring the `join` functionality of `kubeadm` in a way that supports configuring older node versions in a maintainable way.
+* Adjusting the way the skew CI jobs are configured to use `kubeadm` binaries from previous minor versions to configure nodes for those versions.
+
+## Design Details
+
+First, we'll look at the types of changes that actually prevent a specific control plane release from supporting older nodes.
+
+Second, we'll evaluate previous control plane releases, to see how the proposed policy change could have impacted work in past releases.
+
+Finally, we'll list known changes planned for future control plane releases, to evaluate the known impact of this proposal on future work.
+
+### Types of changes
+
+**Feature additions in control plane components that require waiting for node support**
+
+There are two approaches currently used to roll out features that involve changes to both control plane and node components:
+
+1. Enable a feature by default in a given release, require nodes be upgraded to that release 
+   for the feature to work successfully, and expect users to wait until nodes are upgraded to use the new feature.
```

Comment:
There's a slight variation of this in the case of EndpointSlices. Generally we have to enable any new feature in the EndpointSlice controller at least 1 release before kube-proxy reads from it because the EndpointSlice controller needs time to fill out the latest API version. 

For example, EndpointSlice controller needed to write to v1 API for 1 minor version before we updated kube-proxy to read from v1. More recently, KEP #1669 had to wait until KEP #1672 had written the terminating endpoint state for at least 1 minor version before kube-proxy could read from it.

Reply:
I think that offset (wait for a controller to populate data for a release before consumers rely on the data) had more to do with the endpoints → endpointslice transition than the endpointslice v1beta1 → v1 transition, right?

kube-proxy continuing to use endpoints for an additional release didn't cause problems, it just meant we didn't get the benefits of endpointslices as early, right?

Reply:
> More recently, KEP https://github.com/kubernetes/enhancements/issues/1669 had to wait until KEP https://github.com/kubernetes/enhancements/issues/1672 had written the terminating endpoint state for at least 1 minor version before kube-proxy could read from it.

I assume kube-proxy maintained safe / compatible behavior in the absence of terminating endpoint state, right?

Reply:
> As far as I know, yes, but I don't think we have any great version skew test coverage to ensure that.

the ga-only conformance e2e job which would have ensured functionality against endpoints rather than endpointslices, right?

Reply:
>the ga-only conformance e2e job which would have ensured functionality against endpoints rather than endpointslices, right?

Yep, agree that would cover the Endpoints -> EndpointSlice transition. Was less sure about terminating endpoints features because I think that's a bit more nuanced than any of the tests that ga-only job would include.

Reply:
Oh, ok... I may be mixing up two different KEPs conflating v1 graduation of endpointslices and addition of the terminating endpoint info. For #1669, it sounded like the fallback was that kube-proxy behaved as it did before the feature existed:

https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1669-proxy-terminating-endpoints#version-skew-strategy

> Version Skew Strategy
>
> The worse case version skew scenario is that kube-proxy falls back to the existing behavior today where traffic does not fall back to terminating endpoints.

That seems like a sensible, safe, fallback to me.

Reply:
That was the at least the plan (hopefully it was implemented like that).

Reply:
ok, summing up, I think there were two aspects mixed together here:
1. migrating from v1beta1 → v1
2. either waiting one release for a controller to start filling in data, or failing safe when looking at objects a controller may not have filled in yet

1 still works with n-3 skew (switch from v1beta1 to v1 when v1 is released, and you'll work against an n+3 server that stops serving v1beta1)

2 is not impacted by expanding from n-2 to n-3, but it might mean you need to roll out a controller change one version before enabling something in kube-proxy (though I hope https://github.com/kubernetes/enhancements/pull/3920 will let us improve our velocity for things like that)